drivers/adc/adc_sam_afec.c

```diff
@@ -16,143 +20,250 @@
+	afec->AFEC_IER = (1 << data->channel_id);
+
+	/* Start the conversions. */
+	afec->AFEC_CR = AFEC_CR_START;
+}
+
+/**
+ * This is only called once at the beginning of all the conversions,
+ * all channels as a group.
+ */
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_sam_data *data = CONTAINER_OF(ctx, struct adc_sam_data, ctx);
 
-		/* Set start of sample buffer */
-		dev_data->samples[channel].buffer =
-			(u16_t *)seq_tbl->entries[i].buffer;
+	adc_sam_start_conversion(data->dev);
+}
 
-		/* Enable channel */
-		dev_data->active_channels |= BIT(channel);
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct adc_sam_data *data = CONTAINER_OF(ctx, struct adc_sam_data, ctx);
+
+	if (!repeat_sampling) {
+		data->buffer = data->repeat_buffer;
 	}
+}
 
-	/* Enable chosen channels and their interrupts */
-	afec->AFEC_CHER = dev_data->active_channels;
-	afec->AFEC_IER = dev_data->active_channels;
+static int check_buffer_size(const struct adc_sequence *sequence,
+			     u8_t active_channels)
+{
+	size_t needed_buffer_size;
+ 	needed_buffer_size = active_channels * sizeof(u16_t);
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+ 	if (sequence->buffer_size < needed_buffer_size) {
+		SYS_LOG_ERR("Provided buffer is too small (%u/%u)",
+				sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+ 	return 0;
+}
 
-	/* Start conversions */
-	dev_data->measured_channels = 0;
-	dev_data->active_chan_last = dev_data->active_channels;
-	afec->AFEC_CR = AFEC_CR_START;
-	k_sem_take(&dev_data->sem_meas, K_FOREVER);
+static int start_read(struct device *dev, const struct adc_sequence *sequence)
+{
+	struct adc_sam_data *data = DEV_DATA(dev);
+	int error = 0;
+	u32_t channels = sequence->channels;
+
+	data->channels = 0;
+
+	/* Signal an error if the channel selection is invalid (no channels or
+	 * a non-existing one is selected).
+	 */
+	if (channels == 0 ||
+	   (channels & (~0UL << NUM_CHANNELS))) {
+		SYS_LOG_ERR("Invalid selection of channels");
+		return -EINVAL;
+	}
 
-	k_sem_give(&dev_data->mutex_thread);
+	if (sequence->oversampling != 0) {
+		SYS_LOG_ERR("Oversampling is not supported");
+		return -EINVAL;
+	}
 
-	return 0;
+	if (sequence->resolution != 12) {
+		/* TODO JKW: Support the Enhanced Resolution Mode 50.6.3 page
+		 * 1544.
+		 */
+		SYS_LOG_ERR("ADC resolution value %d is not valid",
+			    sequence->resolution);
+		return -EINVAL;
+	}
+
+	if (sequence->options->interval_us != 0) {
+		SYS_LOG_ERR("Spaced intervals not supported");
+		return -EINVAL;
+	}
+
+	u8_t num_active_channels = 0;
+	u8_t channel = 0;
+	while (channels > 0) {
+		if (channels & 1) {
+			++num_active_channels;
+		}
+		channels >>= 1;
+		++channel;
+	}
+
+	error = check_buffer_size(sequence, num_active_channels);
+	if (error) {
+		return error;
+	}
+
+	data->channels = sequence->channels;
```

Comment:
I think you need to move this into `adc_context_start_sampling` to handle repeated samplings.